### ISSUE TITLE
Broken suggestion for lint "hidden lifetime parameters in types"

### ISSUE THREAD

**Issue by lzutao** [rust-lang/rust#65615]
Consider this code:

```rust
macro_rules! hash_impl {
    ($size:ty) => {
        use std::fmt;
        use std::any;

        pub struct Hash { size: $size }

        impl Hash {
            pub fn new(size: $size) -> Self { Self { size } }
        }

        impl fmt::Debug for Hash {
            fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
                f.debug_struct(any::type_name::<Self>()).field("size", &self.size).finish()
            }
        }
    }
}

pub mod foo { hash_impl!(u32); }
pub mod bar { hash_impl!(u64); }
```

When I run `cargo fix --edition-idioms`, I got:

```rust
The following errors were reported:
error: expected one of `!`, `)`, `,`, or `::`, found `<`
  --> src/lib.rs:13:53
   |
13 |             fn fmt(&self, f: &mut fmt::Formatter<'_><'_>) -> Result<(), fmt::Error> {
   |                                                     ^ expected one of `!`, `)`, `,`, or `::` here
...
20 | pub mod foo { hash_impl!(u32); }
   |               ---------------- in this macro invocation

error: expected parameter name, found `'_`
  --> src/lib.rs:13:54
   |
13 |             fn fmt(&self, f: &mut fmt::Formatter<'_><'_>) -> Result<(), fmt::Error> {
   |                                                      ^^ expected parameter name
...
20 | pub mod foo { hash_impl!(u32); }
   |               ---------------- in this macro invocation

error: expected one of `!`, `)`, `,`, or `::`, found `<`
  --> src/lib.rs:13:53
   |
13 |             fn fmt(&self, f: &mut fmt::Formatter<'_><'_>) -> Result<(), fmt::Error> {
   |                                                     ^ expected one of `!`, `)`, `,`, or `::` here
...
21 | pub mod bar { hash_impl!(u64); }
   |               ---------------- in this macro invocation

error: expected parameter name, found `'_`
  --> src/lib.rs:13:54
   |
13 |             fn fmt(&self, f: &mut fmt::Formatter<'_><'_>) -> Result<(), fmt::Error> {
   |                                                      ^^ expected parameter name
...
21 | pub mod bar { hash_impl!(u64); }
   |               ---------------- in this macro invocation

error[E0050]: method `fmt` has 3 parameters but the declaration in trait `std::fmt::Debug::fmt` has 2
  --> src/lib.rs:13:20
   |
13 |             fn fmt(&self, f: &mut fmt::Formatter<'_><'_>) -> Result<(), fmt::Error> {
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 2 parameters, found 3
...
20 | pub mod foo { hash_impl!(u32); }
   |               ---------------- in this macro invocation
   |
   = note: `fmt` from trait: `fn(&Self, &mut std::fmt::Formatter<'_>) -> std::result::Result<(), std::fmt::Error>`

error[E0050]: method `fmt` has 3 parameters but the declaration in trait `std::fmt::Debug::fmt` has 2
  --> src/lib.rs:13:20
   |
13 |             fn fmt(&self, f: &mut fmt::Formatter<'_><'_>) -> Result<(), fmt::Error> {
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected 2 parameters, found 3
...
21 | pub mod bar { hash_impl!(u64); }
   |               ---------------- in this macro invocation
   |
   = note: `fmt` from trait: `fn(&Self, &mut std::fmt::Formatter<'_>) -> std::result::Result<(), std::fmt::Error>`

error: aborting due to 6 previous errors

For more information about this error, try `rustc --explain E0050`.
Original diagnostics will follow.

warning: hidden lifetime parameters in types are deprecated
  --> src/lib.rs:13:35
   |
13 |             fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
   |                                   ^^^^^^^^^^^^^^- help: indicate the anonymous lifetime: `<'_>`
...
20 | pub mod foo { hash_impl!(u32); }
   |               ---------------- in this macro invocation
   |
   = note: `-W elided-lifetimes-in-paths` implied by `-W rust-2018-idioms`

warning: hidden lifetime parameters in types are deprecated
  --> src/lib.rs:13:35
   |
13 |             fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
   |                                   ^^^^^^^^^^^^^^- help: indicate the anonymous lifetime: `<'_>`
...
21 | pub mod bar { hash_impl!(u64); }
   |               ---------------- in this macro invocation

warning: hidden lifetime parameters in types are deprecated
  --> src/lib.rs:13:35
   |
13 |             fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
   |                                   ^^^^^^^^^^^^^^- help: indicate the anonymous lifetime: `<'_>`
...
20 | pub mod foo { hash_impl!(u32); }
   |               ---------------- in this macro invocation
   |
   = note: `-W elided-lifetimes-in-paths` implied by `-W rust-2018-idioms`

warning: hidden lifetime parameters in types are deprecated
  --> src/lib.rs:13:35
   |
13 |             fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
   |                                   ^^^^^^^^^^^^^^- help: indicate the anonymous lifetime: `<'_>`
...
21 | pub mod bar { hash_impl!(u64); }
   |               ---------------- in this macro invocation

    Finished dev [unoptimized + debuginfo] target(s) in 0.35s
```

### Meta

* rustc 1.40.0-nightly (c23a7aa77 2019-10-19)
* cargo 1.40.0-nightly (3a9abe3f0 2019-10-15)

**Comment by lzutao** [comment]
cc @petrochenkov

**Comment by lzutao** [comment]
cc @zackmdavis (you wrote this lint)

**Comment by zackmdavis** [comment]
Thanks for the report. This is a duplicate of #55768.

**Comment by lzutao** [comment]
So is it.
